File: python-backend/indicators/trend_lines.py
"""
Trend Lines v2 - Converted from Pine Script by LonesomeTheBlue
Copy from charting/indicators/trend_lines_v2.py

This indicator finds pivot highs/lows and draws trend lines connecting them.
"""

import logging
from dataclasses import dataclass
from typing import List, Tuple, Optional, Dict, Any

logger = logging.getLogger("indicators.trend_lines")


@dataclass
class TrendLine:
    """Represents a trend line with start and end points"""
    x1: int       # Start bar index
    y1: float     # Start price
    x2: int       # End bar index
    y2: float     # End price
    line_type: str  # 'up' for uptrend (support), 'down' for downtrend (resistance)
    slope: float = 0.0
    strength: int = 2
    age_bars: int = 0


class TrendLinesDetector:
    """
    Trend Lines v2 Indicator
    
    Finds pivot points and draws trend lines connecting them.
    Validates that price doesn't break through the trend line.
    
    Parameters:
    -----------
    pivot_period : int
        Number of bars to look back/forward for pivot detection (default: 20)
    pp_num : int
        Number of pivot points to check for trend lines (default: 3)
    max_lines : int
        Maximum number of trend lines to draw (default: 3)
    """
    
    def __init__(self, pivot_period: int = 20, pp_num: int = 3, max_lines: int = 3):
        self.pivot_period = pivot_period
        self.pp_num = pp_num
        self.max_lines = max_lines
        
        # Storage for pivot values and positions
        self.pivot_highs: List[Tuple[int, float]] = []  # (bar_index, value)
        self.pivot_lows: List[Tuple[int, float]] = []
    
    def _detect_pivot_high(self, highs: List[float], idx: int) -> Optional[float]:
        """
        Detect pivot high at position idx
        A pivot high is the highest point with pivot_period bars on each side being lower
        """
        if idx < self.pivot_period or idx >= len(highs) - self.pivot_period:
            return None
        
        pivot_idx = idx - self.pivot_period
        pivot_val = highs[pivot_idx]
        
        # Check if it's a local maximum
        left_range = highs[pivot_idx - self.pivot_period:pivot_idx]
        right_range = highs[pivot_idx + 1:pivot_idx + self.pivot_period + 1]
        
        if len(left_range) == 0 or len(right_range) == 0:
            return None
            
        if pivot_val >= max(left_range) and pivot_val >= max(right_range):
            return pivot_val
        return None
    
    def _detect_pivot_low(self, lows: List[float], idx: int) -> Optional[float]:
        """
        Detect pivot low at position idx
        A pivot low is the lowest point with pivot_period bars on each side being higher
        """
        if idx < self.pivot_period or idx >= len(lows) - self.pivot_period:
            return None
        
        pivot_idx = idx - self.pivot_period
        pivot_val = lows[pivot_idx]
        
        # Check if it's a local minimum
        left_range = lows[pivot_idx - self.pivot_period:pivot_idx]
        right_range = lows[pivot_idx + 1:pivot_idx + self.pivot_period + 1]
        
        if len(left_range) == 0 or len(right_range) == 0:
            return None
            
        if pivot_val <= min(left_range) and pivot_val <= min(right_range):
            return pivot_val
        return None
    
    def detect(self, candles: List[Dict]) -> Dict[str, Any]:
        """
        Calculate trend lines from OHLCV data
        
        Parameters:
        -----------
        candles : List[Dict]
            List of candles with keys: timestamp, open, high, low, close, volume
            
        Returns:
        --------
        Dict with uptrend_lines, downtrend_lines, filter_type
        """
        highs = [c['high'] for c in candles]
        lows = [c['low'] for c in candles]
        closes = [c['close'] for c in candles]
        n = len(candles)
        
        if n < self.pivot_period * 2 + 10:
            return {
                "uptrend_lines": [],
                "downtrend_lines": [],
                "filter_type": "none",
            }
        
        # Reset pivot storage
        self.pivot_highs = []
        self.pivot_lows = []
        
        # Detect all pivots
        for i in range(self.pivot_period, n):
            ph = self._detect_pivot_high(highs, i)
            if ph is not None:
                pivot_bar = i - self.pivot_period
                self.pivot_highs.append((pivot_bar, ph))
                # Keep only pp_num most recent
                if len(self.pivot_highs) > self.pp_num:
                    self.pivot_highs.pop(0)
            
            pl = self._detect_pivot_low(lows, i)
            if pl is not None:
                pivot_bar = i - self.pivot_period
                self.pivot_lows.append((pivot_bar, pl))
                if len(self.pivot_lows) > self.pp_num:
                    self.pivot_lows.pop(0)
        
        # Find uptrend lines (connecting pivot lows - support)
        uptrend_lines = self._find_uptrend_lines(closes, n)
        
        # Find downtrend lines (connecting pivot highs - resistance)
        downtrend_lines = self._find_downtrend_lines(closes, n)
        
        # Determine filter type
        has_up = len(uptrend_lines) > 0
        has_down = len(downtrend_lines) > 0
        
        if has_up and has_down:
            filter_type = "both"
        elif has_up:
            filter_type = "up"
        elif has_down:
            filter_type = "down"
        else:
            filter_type = "none"
        
        return {
            "uptrend_lines": uptrend_lines,
            "downtrend_lines": downtrend_lines,
            "filter_type": filter_type,
        }
    
    def _find_uptrend_lines(self, closes: List[float], n: int) -> List[TrendLine]:
        """Find valid uptrend lines from pivot lows"""
        lines = []
        
        if len(self.pivot_lows) < 2:
            return lines
        
        # Try pairs of pivot lows (older to newer)
        for i in range(len(self.pivot_lows) - 1):
            for j in range(i + 1, len(self.pivot_lows)):
                pos1, val1 = self.pivot_lows[j]  # More recent (higher position)
                pos2, val2 = self.pivot_lows[i]  # Older (lower position)
                
                # For uptrend, val1 should be > val2 (ascending lows)
                if val1 <= val2:
                    continue
                
                # Calculate slope
                if pos1 == pos2:
                    continue
                    
                slope = (val1 - val2) / (pos1 - pos2)
                
                # Validate: check if price never closes below the line
                valid = True
                for k in range(pos2 + 1, n):
                    line_val = val2 + slope * (k - pos2)
                    if closes[k] < line_val:
                        valid = False
                        break
                
                if valid:
                    # Extend line to current bar
                    end_val = val2 + slope * (n - 1 - pos2)
                    lines.append(TrendLine(
                        x1=pos2,
                        y1=val2,
                        x2=n - 1,
                        y2=end_val,
                        line_type='up',
                        slope=slope,
                        strength=2,
                        age_bars=n - 1 - pos2
                    ))
                    
                    if len(lines) >= self.max_lines:
                        return lines
        
        return lines
    
    def _find_downtrend_lines(self, closes: List[float], n: int) -> List[TrendLine]:
        """Find valid downtrend lines from pivot highs"""
        lines = []
        
        if len(self.pivot_highs) < 2:
            return lines
        
        # Try pairs of pivot highs (older to newer)
        for i in range(len(self.pivot_highs) - 1):
            for j in range(i + 1, len(self.pivot_highs)):
                pos1, val1 = self.pivot_highs[j]  # More recent
                pos2, val2 = self.pivot_highs[i]  # Older
                
                # For downtrend, val1 should be < val2 (descending highs)
                if val1 >= val2:
                    continue
                
                # Calculate slope (negative for downtrend)
                if pos1 == pos2:
                    continue
                    
                slope = (val1 - val2) / (pos1 - pos2)
                
                # Validate: check if price never closes above the line
                valid = True
                for k in range(pos2 + 1, n):
                    line_val = val2 + slope * (k - pos2)
                    if closes[k] > line_val:
                        valid = False
                        break
                
                if valid:
                    # Extend line to current bar
                    end_val = val2 + slope * (n - 1 - pos2)
                    lines.append(TrendLine(
                        x1=pos2,
                        y1=val2,
                        x2=n - 1,
                        y2=end_val,
                        line_type='down',
                        slope=slope,
                        strength=2,
                        age_bars=n - 1 - pos2
                    ))
                    
                    if len(lines) >= self.max_lines:
                        return lines
        
        return lines


def detect_trendlines(
    candles: List[Dict],
    pivot_period: int = 20,
    pp_num: int = 3,
    max_lines: int = 3
) -> Dict[str, Any]:
    """
    Convenience function to detect trend lines.
    
    Returns dict with:
    - uptrend_lines: Support lines (ascending lows, green)
    - downtrend_lines: Resistance lines (descending highs, red)
    - filter_type: 'up', 'down', 'both', or 'none'
    """
    detector = TrendLinesDetector(pivot_period, pp_num, max_lines)
    return detector.detect(candles)

File: src/lib/trendlines.ts
/**
 * 📈 Trend Lines v2 - TypeScript Port
 * 
 * Exact port of the PineScript "Trend Lines v2" by LonesomeTheBlue.
 * This calculates trendlines locally from OHLCV data.
 */

export interface TrendLine {
  type: 'up' | 'down';
  x1: number;
  y1: number;
  x2: number;
  y2: number;
  slope: number;
  strength: number;
  age_bars: number;
}

export interface Candle {
  time: number;
  open: number;
  high: number;
  low: number;
  close: number;
}

export interface TrendLinesResult {
  uptrend_lines: TrendLine[];
  downtrend_lines: TrendLine[];
  filter_type: 'up' | 'down' | 'both' | 'none';
}

/**
 * Detect pivot high at position idx
 * A pivot high is the highest point with pivotPeriod bars on each side being lower
 */
function detectPivotHigh(highs: number[], idx: number, pivotPeriod: number): number | null {
  if (idx < pivotPeriod || idx >= highs.length - pivotPeriod) {
    return null;
  }
  
  const pivotIdx = idx - pivotPeriod;
  const pivotVal = highs[pivotIdx];
  
  // Check if it's a local maximum
  const leftRange = highs.slice(pivotIdx - pivotPeriod, pivotIdx);
  const rightRange = highs.slice(pivotIdx + 1, pivotIdx + pivotPeriod + 1);
  
  if (leftRange.length === 0 || rightRange.length === 0) {
    return null;
  }
  
  const leftMax = Math.max(...leftRange);
  const rightMax = Math.max(...rightRange);
  
  if (pivotVal >= leftMax && pivotVal >= rightMax) {
    return pivotVal;
  }
  return null;
}

/**
 * Detect pivot low at position idx
 * A pivot low is the lowest point with pivotPeriod bars on each side being higher
 */
function detectPivotLow(lows: number[], idx: number, pivotPeriod: number): number | null {
  if (idx < pivotPeriod || idx >= lows.length - pivotPeriod) {
    return null;
  }
  
  const pivotIdx = idx - pivotPeriod;
  const pivotVal = lows[pivotIdx];
  
  // Check if it's a local minimum
  const leftRange = lows.slice(pivotIdx - pivotPeriod, pivotIdx);
  const rightRange = lows.slice(pivotIdx + 1, pivotIdx + pivotPeriod + 1);
  
  if (leftRange.length === 0 || rightRange.length === 0) {
    return null;
  }
  
  const leftMin = Math.min(...leftRange);
  const rightMin = Math.min(...rightRange);
  
  if (pivotVal <= leftMin && pivotVal <= rightMin) {
    return pivotVal;
  }
  return null;
}

/**
 * Find valid uptrend lines from pivot lows
 */
function findUptrendLines(
  pivotLows: [number, number][],
  closes: number[],
  n: number,
  maxLines: number
): TrendLine[] {
  const lines: TrendLine[] = [];
  
  if (pivotLows.length < 2) {
    return lines;
  }
  
  // Try pairs of pivot lows (older to newer)
  for (let i = 0; i < pivotLows.length - 1; i++) {
    for (let j = i + 1; j < pivotLows.length; j++) {
      const [pos1, val1] = pivotLows[j]; // More recent (higher position)
      const [pos2, val2] = pivotLows[i]; // Older (lower position)
      
      // For uptrend, val1 should be > val2 (ascending lows)
      if (val1 <= val2) {
        continue;
      }
      
      // Calculate slope
      if (pos1 === pos2) {
        continue;
      }
      
      const slope = (val1 - val2) / (pos1 - pos2);
      
      // Validate: check if price never closes below the line
      let valid = true;
      for (let k = pos2 + 1; k < n; k++) {
        const lineVal = val2 + slope * (k - pos2);
        if (closes[k] < lineVal) {
          valid = false;
          break;
        }
      }
      
      if (valid) {
        // Extend line to current bar
        const endVal = val2 + slope * (n - 1 - pos2);
        lines.push({
          type: 'up',
          x1: pos2,
          y1: val2,
          x2: n - 1,
          y2: endVal,
          slope: slope,
          strength: 2,
          age_bars: n - 1 - pos2
        });
        
        if (lines.length >= maxLines) {
          return lines;
        }
      }
    }
  }
  
  return lines;
}

/**
 * Find valid downtrend lines from pivot highs
 */
function findDowntrendLines(
  pivotHighs: [number, number][],
  closes: number[],
  n: number,
  maxLines: number
): TrendLine[] {
  const lines: TrendLine[] = [];
  
  if (pivotHighs.length < 2) {
    return lines;
  }
  
  // Try pairs of pivot highs (older to newer)
  for (let i = 0; i < pivotHighs.length - 1; i++) {
    for (let j = i + 1; j < pivotHighs.length; j++) {
      const [pos1, val1] = pivotHighs[j]; // More recent
      const [pos2, val2] = pivotHighs[i]; // Older
      
      // For downtrend, val1 should be < val2 (descending highs)
      if (val1 >= val2) {
        continue;
      }
      
      // Calculate slope (negative for downtrend)
      if (pos1 === pos2) {
        continue;
      }
      
      const slope = (val1 - val2) / (pos1 - pos2);
      
      // Validate: check if price never closes above the line
      let valid = true;
      for (let k = pos2 + 1; k < n; k++) {
        const lineVal = val2 + slope * (k - pos2);
        if (closes[k] > lineVal) {
          valid = false;
          break;
        }
      }
      
      if (valid) {
        // Extend line to current bar
        const endVal = val2 + slope * (n - 1 - pos2);
        lines.push({
          type: 'down',
          x1: pos2,
          y1: val2,
          x2: n - 1,
          y2: endVal,
          slope: slope,
          strength: 2,
          age_bars: n - 1 - pos2
        });
        
        if (lines.length >= maxLines) {
          return lines;
        }
      }
    }
  }
  
  return lines;
}

/**
 * Calculate trend lines from OHLCV candles
 * 
 * @param candles - Array of candles sorted by time (oldest first)
 * @param pivotPeriod - Number of bars to look back/forward for pivot detection (default: 20)
 * @param ppNum - Number of pivot points to keep (default: 3)
 * @param maxLines - Maximum number of trend lines per direction (default: 3)
 */
export function calculateTrendLines(
  candles: Candle[],
  pivotPeriod: number = 20,
  ppNum: number = 3,
  maxLines: number = 3
): TrendLinesResult {
  const n = candles.length;
  
  if (n < pivotPeriod * 2 + 10) {
    return {
      uptrend_lines: [],
      downtrend_lines: [],
      filter_type: 'none'
    };
  }
  
  const highs = candles.map(c => c.high);
  const lows = candles.map(c => c.low);
  const closes = candles.map(c => c.close);
  
  // Storage for pivot values and positions
  const pivotHighs: [number, number][] = []; // [bar_index, value]
  const pivotLows: [number, number][] = [];
  
  // Detect all pivots
  for (let i = pivotPeriod; i < n; i++) {
    const ph = detectPivotHigh(highs, i, pivotPeriod);
    if (ph !== null) {
      const pivotBar = i - pivotPeriod;
      pivotHighs.push([pivotBar, ph]);
      // Keep only ppNum most recent
      if (pivotHighs.length > ppNum) {
        pivotHighs.shift();
      }
    }
    
    const pl = detectPivotLow(lows, i, pivotPeriod);
    if (pl !== null) {
      const pivotBar = i - pivotPeriod;
      pivotLows.push([pivotBar, pl]);
      if (pivotLows.length > ppNum) {
        pivotLows.shift();
      }
    }
  }
  
  // Find uptrend lines (connecting pivot lows - support)
  const uptrendLines = findUptrendLines(pivotLows, closes, n, maxLines);
  
  // Find downtrend lines (connecting pivot highs - resistance)
  const downtrendLines = findDowntrendLines(pivotHighs, closes, n, maxLines);
  
  // Determine filter type
  const hasUp = uptrendLines.length > 0;
  const hasDown = downtrendLines.length > 0;
  
  let filterType: 'up' | 'down' | 'both' | 'none';
  if (hasUp && hasDown) {
    filterType = 'both';
  } else if (hasUp) {
    filterType = 'up';
  } else if (hasDown) {
    filterType = 'down';
  } else {
    filterType = 'none';
  }
  
  return {
    uptrend_lines: uptrendLines,
    downtrend_lines: downtrendLines,
    filter_type: filterType
  };
}
